proposals/0000-package-internals.rst

```diff
@@ -0,0 +1,148 @@
+Accessing package internal modules
+==================================
+
+.. author:: Richard Eisenberg and Simon Peyton Jones
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/528>`_.
+.. sectnum::
+.. contents::
+
+Package authors include an ``exposed-modules`` stanza in ``.cabal`` files to
+list the modules available for import outside the package. Other modules are
+not available for import. Yet it is sometimes useful to access even unexposed
+("internal") modules. This proposal offers a way to do so.
+
+Note that internal modules of a package are *not* a part of that package's
+published API and thus changes therein would *not* require a major version
+bump. These modules do remain internal!
+
+Motivation
+----------
+
+1. Suppose package ``pack`` has several internal modules, whose exports are
+   not intended for the public API of ``pack``. Yet the author of ``pack``
+   still wants to test functions exported by these internal modules.
+
+   The author currently has two options, both bad:
+
+   1. Expose the internal modules, perhaps naming them with ``Internal`` or
+      similar.
+
+   2. Structure the project to recompile the entire package in order to run
+      the testsuite, instead of using the already-compiled package.
```

Comment:
The blessed way is to put these modules into an internal library, visible for test suite, but invisible for general public. For a real world example see https://hackage.haskell.org/package/attoparsec-0.14.4/attoparsec.cabal

Reply:
It is not raised in Richard's motivations, but private libraries, just as well as unexposed modules, run the risk of exposing functions without exposing enough type constructors to say the functions' types. So if we make private libraries, I will actually want a way to depend on them anyway, much in the manner that is proposed here.